invokeai/app/api/routers/models.py

```diff
@@ -54,6 +55,90 @@ async def list_models() -> ModelsList:
     models = parse_obj_as(ModelsList, { "models": models_raw })
     return models
 
+# Kent wrote the below code. It is highly suspect, and should be reviewed. Once all issues have been identified and fixed, this comment can be removed.
+# Seriously, check this. I have no idea how to test it.
+
+""" Add Model """
+@models_router.post(
+    "/",
+    operation_id="add_model",
+    responses={201: {"model": Union[CkptModelInfo, DiffusersModelInfo], "new_model_list": ModelsList}},
+)
+async def add_model(
+    model_info: Union[CkptModelInfo, DiffusersModelInfo],
+) -> Union[CkptModelInfo, DiffusersModelInfo]:
+    """Adds a new model"""
+    try:
+        model_name = model_info["model_name"]
+        del model_info["model_name"]
```

Comment:
This is invalid. You shouldn't add `model_name` to `ModelInfo` just to be able to pass it in. Instead, define a new request class that has the model name and model info. Something like:

```py
class CreateModelRequest(BaseModel):
    name: str = Field(description="The name of the model")
    info: Annotated[Union[(CkptModelInfo,DiffusersModelInfo)], Field(discriminator="format")] = Field(description="The model info")
```

(Note: I think you combine the two `Field`s above... add the description inside the annotation)

Reply:
If the Model Info doesn't have the Name, how is the Model List info going to have references for it? Is the Name something passed around as a label for the entirety of the `info` object even though it's not explicitly defined?

Additionally, what ID is used to select in order to update the model if not name?

Reply:
So - Walking through my thinking to figure out where assumptions are off.

Name seemed to be needed as part of the "Model" object generally. (I didn't see where else it would be defined in order to reference it)

The old/legacy code did the same function as above - Explicitly, for this function, splitting the `info` object into a name and attributes, as the backend code (in model_manager) is specifically looking for the request to be formatted like (model_name, model_attributes, clobber) as far as I understand it.

Recognizing that I could create another class for the request, but also noting that the Name is needed elsewhere (presumably it should be included in the models list?) I figured adding it to the core model info definition would be ideal.

Ok - now I can be educated. :)

Reply:
My read is that the models are stored in a dictionary (or dict-like) structure, where the name is the key and the info is the value. The big issue is this:
```py
        model_name = model_info["model_name"]
        del model_info["model_name"]
```
That would delete the `model_name` property from the `model_info` object, which would make it invalid (since the `model_name` is a required property the way you've added it), and pydantic would throw a fit over it eventually.

Reply:
Ahhh. I see. Ok - will need to update this tonight!